src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,386 @@
+        throws InvalidKeyException, NoSuchAlgorithmException {
+
+        if (salt == null) {
+            salt = new byte[hmacLen];
+        }
+        Mac hmacObj = Mac.getInstance(hmacAlgName);
+        hmacObj.init(new SecretKeySpec(salt, hmacAlgName));
+
+        if (inputKeyMaterial == null) {
+            return hmacObj.doFinal();
+        } else {
+            return hmacObj.doFinal(inputKeyMaterial);
+        }
+    }
+
+    /**
+     * Perform the HKDF-Expand operation.
+     *
+     * @param prk
+     *     the pseudorandom key used for HKDF-Expand
+     * @param info
+     *     optional context and application specific information or {@code null}
+     *     if no info data is provided.
+     * @param outLen
+     *     the length in bytes of the required output
+     *
+     * @return a byte array containing the complete KDF output.  This will be at
+     *     least as long as the requested length in the {@code outLen}
+     *     parameter, but will be rounded up to the nearest multiple of the HMAC
+     *     output length.
+     *
+     * @throws InvalidKeyException
+     *     if an invalid PRK was provided through the {@code HKDFParameterSpec}
+     *     or derived during the extract phase.
+     */
+    private byte[] hkdfExpand(byte[] prk, byte[] info, int outLen)
+        throws InvalidKeyException, NoSuchAlgorithmException {
+        byte[] kdfOutput;
+        SecretKey pseudoRandomKey = new SecretKeySpec(prk, hmacAlgName);
+
+        Mac hmacObj = Mac.getInstance(hmacAlgName);
+
+        // Calculate the number of rounds of HMAC that are needed to
+        // meet the requested data.  Then set up the buffers we will need.
+        if (CipherCore.getKeyBytes(pseudoRandomKey).length < hmacLen) {
+            throw new InvalidKeyException(
+                "prk must be at least " + hmacLen + " bytes");
+        }
+        hmacObj.init(pseudoRandomKey);
+        if (info == null) {
+            info = new byte[0];
+        }
+        int rounds = (outLen + hmacLen - 1) / hmacLen;
+        kdfOutput = new byte[rounds * hmacLen];
+
+        for (int i = 0, offset = 0; i < rounds; i++, offset += hmacLen) {
+            // Calculate this round
+            try {
+                if (i > 0) {
+                    hmacObj.update(kdfOutput, Math.max(0, offset - hmacLen),
+                                   hmacLen); // add T(i-1)
+                }
+                hmacObj.update(info); // Add info
+                hmacObj.update((byte) (i + 1)); // Add round number
+                hmacObj.doFinal(kdfOutput, offset);
+            } catch (ShortBufferException sbe) {
+                // This really shouldn't happen given that we've
+                // sized the buffers to their largest possible size up-front,
+                // but just in case...
+                throw new RuntimeException(sbe);
+            }
+        }
+
+        return Arrays.copyOf(kdfOutput, outLen);
```

Comment:
Here is an alternative solution which does not need `Arrays.copyOf()` call:
```suggestion
        kdfOutput = new byte[outLen];
        int i = 0;
        int offset = 0;
        try {
            while (i < rounds) {
                if (i > 0) {
                    hmacObj.update(kdfOutput, offset - hmacLen, hmacLen); // add T(i-1)
                }
                hmacObj.update(info);                   // Add info
                hmacObj.update((byte) ++i);             // Add round number
                if (i == rounds && (outLen - offset < hmacLen)) {
                    // special handling for last chunk
                    byte[] tmp = hmacObj.doFinal();
                    System.arraycopy(tmp, 0, kdfOutput, offset,
                            outLen - offset);
                    offset = outLen;
                } else {
                    hmacObj.doFinal(kdfOutput, offset);
                    offset += hmacLen;
                }
            }
        } catch (ShortBufferException sbe) {
            // This really shouldn't happen given that we've
            // sized the buffers to their largest possible size up-front,
            // but just in case...
            throw new RuntimeException(sbe);
        }
        return kdfOutput;
```